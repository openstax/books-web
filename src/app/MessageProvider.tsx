import { shouldPolyfill } from '@formatjs/intl-pluralrules/should-polyfill';
import React from 'react';
import { IntlProvider } from 'react-intl';
import enMessages from './messages/en';

// https://formatjs.io/docs/polyfills/intl-pluralrules/#dynamic-import--capability-detection
async function polyfill(locale: 'en') {
  if (shouldPolyfill()) {
    await import('@formatjs/intl-pluralrules/polyfill');
  }

  // boolean added by the polyfill
  if ((Intl.PluralRules as (typeof Intl.PluralRules & {polyfilled?: boolean})).polyfilled) {
    switch (locale) {
      case 'en':
        await import('@formatjs/intl-pluralrules/locale-data/en');
        break;
      default:
        throw new Error('unsupported locale in messageProvider');
    }
  }
}

polyfill('en');

// tslint:disable-next-line:variable-name
const MessageProvider: React.SFC<{onError?: () => void}> = (props) =>
  <IntlProvider onError={props.onError} locale='en' messages={enMessages}>{props.children}</IntlProvider>;

export default MessageProvider;
