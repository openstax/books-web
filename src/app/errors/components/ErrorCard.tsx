import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components';
import { ExclamationTriangle } from 'styled-icons/fa-solid/ExclamationTriangle';
import { textStyle } from '../../components/Typography/base';
import theme from '../../theme';

const margin = 3.0;

// tslint:disable-next-line:variable-name
const Card = styled.div`
  display: flex;
  flex-direction: column;
  margin: auto;
  overflow: hidden;
  padding: ${margin}rem;
  width: 30rem;
  ${textStyle};
  background-color: white;
  box-shadow: 0 0 2rem rgba(0, 0, 0, 0.05), 0 0 4rem rgba(0, 0, 0, 0.08);
  border-radius: 0.5rem;
`;

// tslint:disable-next-line:variable-name
const Header = styled.header`
  display: flex;
  align-items: center;
  margin-bottom: ${margin * 0.5}rem;
`;

// tslint:disable-next-line:variable-name
const Heading = styled.h1`
  display: flex;
  align-items: center;
  margin: 0;
  font-size: 2.4rem;
  font-weight: bold;
  text-align: center;
`;

// tslint:disable-next-line:variable-name
const Body = styled.div`
  display: flex;
  flex-direction: column;
  font-size: 1.4rem;
  margin-top: ${margin * 0.5}rem;
  margin-bottom: ${margin * 0.5}rem;

  ${Card.Header} + & { /* stylelint-disable */
  margin-top: 0;
  }
`;

// tslint:disable-next-line:variable-name
export const Footer = styled.div`
  display: flex;
  justify-content: flex-end;
`;

// tslint:disable-next-line:variable-name
const ErrorPanel = styled.div`
  color: white;
  padding: 1rem;
  margin-top: 2rem;
  font-size: 1.2rem;
  background-color: ${theme.color.secondary.lightGray.base};
`;

interface Props {
  error: Error;
  className?: string;
  footer?: JSX.Element;
}

// tslint:disable-next-line:variable-name
const ErrorCard: React.SFC<Props> = ({
  error, className, footer,
}) => {

  return (
    <Card className={className}>
      <Header>
        <FormattedMessage id='i18n:error:boundary:heading'>
          {(message) => (
            <Heading>
              <ExclamationTriangle height='45px' color='red' /> {message}
            </Heading>
          )}
        </FormattedMessage>
      </Header>
      <Body>
        <FormattedMessage id='i18n:error:boundary:body'>
          {(body) => (<div>{body}</div>)}
        </FormattedMessage>
        <ErrorPanel>
          {error.toString()}
        </ErrorPanel>
        <FormattedMessage id='i18n:error:boundary:action'>
          {(action) => (<p>{action}</p>)}
        </FormattedMessage>
      </Body>
      {footer}
    </Card>
  );
};

export default ErrorCard;
