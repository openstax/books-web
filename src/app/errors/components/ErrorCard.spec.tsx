import React from 'react';
import renderer from 'react-test-renderer';
import MessageProvider from '../../MessageProvider';
import ErrorCard from './ErrorCard';

describe('ErrorCard', () => {

  it('matches snapshot', () => {
    const tree = renderer
      .create(<MessageProvider><ErrorCard /></MessageProvider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
