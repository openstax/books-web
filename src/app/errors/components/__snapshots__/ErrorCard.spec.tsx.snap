// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorCard matches snapshot 1`] = `
.c6 {
  font-weight: 300;
  padding: 1.5rem 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: auto;
  overflow: hidden;
  width: 40rem;
  font-family: Neue Helvetica W01;
  color: #424242;
  background-color: white;
  box-shadow: 0 0 2rem rgba(0,0,0,0.05),0 0 4rem rgba(0,0,0,0.08);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 1.5rem;
  padding: 1.5rem 3rem;
  background: #f1f1f1;
  border-bottom: solid 0.1rem #fafafa;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.4rem;
  line-height: 1.6rem;
  font-weight: normal;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0;
  font-weight: 500;
  padding: 1.5rem 0;
}

.c5 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 2.4rem;
  line-height: 3rem;
  -webkit-letter-spacing: -0.02rem;
  -moz-letter-spacing: -0.02rem;
  -ms-letter-spacing: -0.02rem;
  letter-spacing: -0.02rem;
  padding: 1.5rem 0 1rem 0;
  margin: 0;
  font-weight: 400;
  padding: 1.5rem 0;
}

.c4 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.4rem;
  line-height: 1.6rem;
  font-weight: normal;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0 3rem;
}

+ .c4 {
  margin-top: 0;
}

.c3 {
  height: 2.5rem;
  width: 2.5rem;
  padding: 0.4rem;
  margin-right: 0;
  padding-right: 0;
  color: #818181;
  height: 2rem;
  width: 2rem;
}

.c3:hover {
  color: #5e6062;
}

@media screen and (max-width:75em) {
  .c5 {
    font-size: 1.6rem;
    line-height: 2rem;
  }
}

<div
  className="c0"
>
  <header
    className="c1"
  >
    <h1
      className="c2"
    >
      Error
    </h1>
    <svg
      aria-hidden="true"
      className="c3"
      focusable="false"
      height="15px"
      version="1.1"
      viewBox="0 0 15 15"
      width="15px"
    >
      <g
        fill="none"
        fillRule="evenodd"
        stroke="none"
        strokeWidth="1"
      >
        <g
          fill="currentColor"
          transform="translate(-302.000000, -18.000000)"
        >
          <g
            transform="translate(302.000000, 18.000000)"
          >
            <path
              d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
            />
          </g>
        </g>
      </g>
    </svg>
  </header>
  <div
    className="c4"
  >
    <h3
      className="c5"
    >
      Uh-oh, there's been a glitch
    </h3>
    <div
      className="c6"
      dangerouslySetInnerHTML={
        Object {
          "__html": "
  Weâ€™re not quite sure what went wrong. You can try refreshing the page, or contact
  our <a href=\\"https://openstax.secure.force.com/help\\" target=\\"_blank\\">Support Center</a> for help.
",
        }
      }
    />
  </div>
</div>
`;
